File: aspnet/src/Persistence/TwoStepWriteRepository.cs
using Npgsql;
using static src.Observability.EventLog;

namespace src.Persistence;

public interface ITwoStepWriteRepository
{
    Task WriteTwoStepsAsync(string req, Guid reqId, int gapMs, bool txnMode, CancellationToken ct);
}

public class TwoStepWriteRepository(NpgsqlDataSource dataSource) : ITwoStepWriteRepository
{
    public async Task WriteTwoStepsAsync(string req, Guid reqId, int gapMs, bool txnMode, CancellationToken ct)
    {
        await using var conn = await dataSource.OpenConnectionAsync(ct);

        if (txnMode)
        {
            await using var txn = await conn.BeginTransactionAsync(ct);
            await InsertStep(conn, txn, reqId, 1, req, ct);
            Log(req, Stage.Db, "gap_started", $"gap_ms={gapMs} tx=true");
            await Task.Delay(gapMs, ct);
            Log(req, Stage.Db, "gap_completed");
            await InsertStep(conn, txn, reqId, 2, req, ct);
            await txn.CommitAsync(ct);
            Log(req, Stage.Db, "txn_committed");
        }
        else
        {
            await InsertStep(conn, null, reqId, 1, req, ct);
            Log(req, Stage.Db, "gap_started", $"gap_ms={gapMs} tx=false");
            await Task.Delay(gapMs, ct);
            Log(req, Stage.Db, "gap_completed");
            await InsertStep(conn, null, reqId, 2, req, ct);
        }
    }

    private static async Task InsertStep(NpgsqlConnection conn, NpgsqlTransaction? txn,
                                         Guid reqId, int step, string req, CancellationToken ct)
    {
        Log(req, Stage.Db, "insert_started", $"step={step}");
        await using var cmd = new NpgsqlCommand(
            $"INSERT INTO txn_writes(req_id, step) VALUES(@reqId, @step) /* req={reqId} */", conn, txn);
        cmd.Parameters.AddWithValue("reqId", reqId);
        cmd.Parameters.AddWithValue("step", step);
        await cmd.ExecuteNonQueryAsync(ct);
        Log(req, Stage.Db, "insert_completed", $"step={step}");
    }
}

File: spring-mvc/src/main/java/com/example/spring/persistence/TwoStepWriteRepository.java
package com.example.spring.persistence;

import com.example.spring.cancellation.CancellationSource;
import com.example.spring.cancellation.CancelledException;
import com.example.spring.observability.EventLog;
import jakarta.annotation.PostConstruct;
import org.springframework.stereotype.Repository;

import javax.sql.DataSource;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.sql.Statement;

public interface TwoStepWriteRepository {
    void writeTwoSteps(String req, String reqId, int gapMs, boolean txnMode, CancellationSource source);
}

@Repository
class TwoStepWriteRepositoryImpl implements TwoStepWriteRepository {

    private final DataSource dataSource;

    TwoStepWriteRepositoryImpl(DataSource dataSource) {
        this.dataSource = dataSource;
    }

    @PostConstruct
    void createTableIfNotExists() {
        try (Connection conn = dataSource.getConnection();
             Statement stmt = conn.createStatement()) {
            stmt.execute("CREATE TABLE IF NOT EXISTS txn_writes " +
                "(req_id UUID NOT NULL, step INT NOT NULL, ts TIMESTAMPTZ NOT NULL DEFAULT NOW(), " +
                "PRIMARY KEY(req_id, step))");
        } catch (SQLException e) {
            throw new RuntimeException("Failed to create txn_writes table", e);
        }
    }

    @Override
    public void writeTwoSteps(String req, String reqId, int gapMs, boolean txnMode, CancellationSource source) {
        EventLog.log(req, EventLog.Stage.DB, "two_step_started", "gap_ms=" + gapMs + " tx=" + txnMode);
        try (Connection conn = dataSource.getConnection()) {
            if (txnMode) conn.setAutoCommit(false);
            try {
                insertStep(conn, reqId, 1, req, source);

                EventLog.log(req, EventLog.Stage.DB, "gap_started", "gap_ms=" + gapMs);
                try {
                    Thread.sleep(gapMs);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    EventLog.log(req, EventLog.Stage.DB, "cancellation_detected", "phase=gap");
                    if (txnMode) conn.rollback();
                    throw new CancelledException();
                }
                EventLog.log(req, EventLog.Stage.DB, "gap_completed");

                insertStep(conn, reqId, 2, req, source);

                if (txnMode) {
                    conn.commit();
                    EventLog.log(req, EventLog.Stage.DB, "txn_committed");
                }
            } catch (SQLException e) {
                if (txnMode) { try { conn.rollback(); } catch (SQLException ignore) {} }
                if (source.isCancelled() || "57014".equals(e.getSQLState())) {
                    EventLog.log(req, EventLog.Stage.DB, "cancellation_detected", "phase=insert");
                    throw new CancelledException();
                }
                throw new RuntimeException(e);
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    private void insertStep(Connection conn, String reqId, int step, String req, CancellationSource source)
            throws SQLException {
        EventLog.log(req, EventLog.Stage.DB, "insert_started", "step=" + step);
        try (PreparedStatement pstmt = conn.prepareStatement(
                "INSERT INTO txn_writes(req_id, step) VALUES (?::uuid, ?) /* req=" + reqId + " */")) {
            pstmt.setString(1, reqId);
            pstmt.setInt(2, step);
            source.setActiveStatement(pstmt);
            try {
                pstmt.execute();
                EventLog.log(req, EventLog.Stage.DB, "insert_completed", "step=" + step);
            } finally {
                source.clearActiveStatement();
            }
        }
    }
}
